public/service/input_auto_retry_test.go

```diff
@@ -339,3 +339,51 @@ func TestAutoRetryBuffer(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, exp3, string(b))
 }
+
+func TestAutoRetryReadAfterClose(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+
+	readerImpl := newMockInput()
+	readerImpl.msgsToSnd = append(readerImpl.msgsToSnd, NewMessage([]byte("foo")))
+
+	pres := AutoRetryNacks(readerImpl)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case readerImpl.connChan <- nil:
+		case <-time.After(time.Second):
+			t.Error("Timed out")
+		}
+		select {
+		case readerImpl.readChan <- nil:
+		case <-time.After(time.Second):
+			t.Error("Timed out")
+		}
+
+		// Force mockInput.Read() to return ErrEndOfInput
+		close(readerImpl.readChan)
+	}()
+
+	require.NoError(t, pres.Connect(ctx))
+
+	msg, _, err := pres.Read(ctx)
```

Comment:
It's because this read is never acked, so as far as the auto retryer is concerned we're not finished, as if this message gets nacked we need to ensure the input is still being read from.

Reply:
Ahhhhh! That makes sense. I forgot about the ack function... Indeed, it has to be called to unblock the workflow. Thanks a million! 😘